### PR TITLE
s3 persistence (atum, sdk fs usage, ...)

### PR DESCRIPTION
**Pull request by dk1844**
Implementation of Enceladus on S3 wave 2:
 - includes Atum for s3 v3.0.0
 - `FileSystemVersionUtils` renamed to `HdfsUtils` and then direct `HdfsUtils` was replaced by a trait `DistributedFsUtils` backed by `HdfsUtils` and `S3FsUtils implementations.
   -AWS S3 SDK ised for the S3FsUtils implementations (measurements, cleanup, guards) and unit tests for it
 - scalaTest version updated, mockito-scala added
 - the non-splittable-to-temporary splittable conversion is not used for S3.

The S3 SDK usage relies on config values being present:
 - `s3.region`
 - `s3.kmsKeyId` (expected to be supplied externally as an override)

`HdfsUtils` is used directly (not via the trait and substitutable by `S3FsUtils`) in HyperConformance and in MenasCredentials (this can be discussed if this is desired or not).

## Test run and its effect of raw vs post-std-post-conformance:

```bash
$ diff <(cat raw_info3 | jq ".") <(cat publish_info3 | jq ".")
10c10,60
<     "additionalInfo": {}
---
>     "additionalInfo": {
>       "conform_size_ratio": "98.86 %",
>       "std_yarn_deploy_mode": "client",
>       "conform_username": "user",
>       "conform_input_data_size": "45330",
>       "raw_format": "csv",
>       "conform_output_data_size": "47020",
>       "std_application_id": "application_1590755537141_0258",
>       "source_record_count": "223929",
>       "std_spark_master": "yarn",
>       "conform_output_dir_size": "47020",
>       "conform_records_failed": "131",
>       "std_executor_memory": "4743M",
>       "std_executors_num": "2",
>       "conform_errors_count": "131",
>       "std_username": "user",
>       "conform_driver_memory": "2048M",
>       "std_output_data_size": "45330",
>       "std_records_succeeded": "512",
>       "std_records_failed": "131",
>       "std_executor_cores": "2",
>       "conform_enceladus_version": "2.12.0-AWS-SNAPSHOT",
>       "conform_spark_master": "yarn",
>       "std_errors_count": "131",
>       "conform_executor_memory": "4743M",
>       "std_output_dir_size": "47337",
>       "std_input_dir_size": "82922",
>       "conform_executors_num": "2",
>       "std_output_dir": "s3://MY_BUCKET/superhero2/std/standardized-dk_test1_emr285-2-2020-08-06-1",
>       "conform_input_dir": "s3://MY_BUCKET/superhero2/std/standardized-dk_test1_emr285-2-2020-08-06-1",
>       "enceladus_info_version": "1",
>       "std_input_dir": "s3://MY_BUCKET/superhero/raw/2020/08/06/v1",
>       "std_enceladus_version": "2.12.0-AWS-SNAPSHOT",
>       "conform_records_succeeded": "512",
>       "conform_data_size_ratio": "98.86 %",
>       "conform_cmd_line_args": "--menas-credentials-file /home/ec2-user/enceladusPoc/menas-credentials.properties --dataset-name dk_test1_emr285 --dataset-version 2 --report-date 2020-08-06 --report-version 1",
>       "std_input_data_size": "81865",
>       "std_cmd_line_args": "--menas-credentials-file /home/ec2-user/enceladusPoc/menas-credentials.properties --dataset-name dk_test1_emr285 --dataset-version 2 --report-date 2020-08-06 --report-version 1 --raw-format csv --header true",
>       "conform_record_count": "643",
>       "conform_executor_cores": "2",
>       "conform_input_dir_size": "47564",
>       "std_driver_memory": "2048M",
>       "std_size_ratio": "54.67 %",
>       "conform_yarn_deploy_mode": "client",
>       "std_record_count": "643",
>       "conform_output_dir": "s3://MY_BUCKET/superhero2/publish/enceladus_info_date=2020-08-06/enceladus_info_version=1",
>       "conform_application_id": "application_1590755537141_0259",
>       "raw_record_count": "223929",
>       "enceladus_info_date": "2020-08-06",
>       "std_data_size_ratio": "54.67 %"
>     }
11a62
>   "runUniqueId": "3270723c-5a80-43b1-ab0a-cb206b6e1559",
22c73
<           "controlType": "controlType.count",
---
>           "controlType": "count",
24c75
<           "controlValue": 223929
---
>           "controlValue": "223929"
37c88
<           "controlType": "controlType.count",
---
>           "controlType": "count",
39c90,141
<           "controlValue": 223929
---
>           "controlValue": "223929"
>         }
>       ]
>     },
>     {
>       "name": "Standardization - End",
>       "software": "Atum",
>       "version": "0.3.0-SNAPSHOT",
>       "processStartTime": "18-09-2020 10:03:49 +0000",
>       "processEndTime": "18-09-2020 10:04:01 +0000",
>       "workflowName": "Standardization",
>       "order": 3,
>       "controls": [
>         {
>           "controlName": "recordcount",
>           "controlType": "count",
>           "controlCol": "*",
>           "controlValue": "643"
>         }
>       ]
>     },
>     {
>       "name": "Conformance - Start",
>       "software": "Atum",
>       "version": "0.3.0-SNAPSHOT",
>       "processStartTime": "18-09-2020 10:09:56 +0000",
>       "processEndTime": "18-09-2020 10:10:06 +0000",
>       "workflowName": "Conformance",
>       "order": 4,
>       "controls": [
>         {
>           "controlName": "recordcount",
>           "controlType": "count",
>           "controlCol": "*",
>           "controlValue": "643"
>         }
>       ]
>     },
>     {
>       "name": "Conformance - End",
>       "software": "Atum",
>       "version": "0.3.0-SNAPSHOT",
>       "processStartTime": "18-09-2020 10:10:06 +0000",
>       "processEndTime": "18-09-2020 10:10:08 +0000",
>       "workflowName": "Conformance",
>       "order": 5,
>       "controls": [
>         {
>           "controlName": "recordcount",
>           "controlType": "count",
>           "controlCol": "*",
>           "controlValue": "643"
```
(bucket name is redacted and replaced by `MY_BUCKET`)

Sorry for the PR being that large.

### CODE REVIEW

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala [hunk 1]
```diff
@@ -51,6 +50,8 @@ trait StreamingFixture extends FunSuite with SparkTestBase with MockitoSugar {
     when(configStub.containsKey(menasUriKey)).thenReturn(true)
     when(configStub.getString(menasUriKey)).thenReturn("https://mymenas.org")
     when(configStub.containsKey(menasAuthKeytabKey)).thenReturn(true)
+    when(configStub.containsKey(menasCredentialsFileKey)).thenReturn(false)
```
Comment (dk1844) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:53]: explicit stub definition instead of relying on the false-returning default behavior.

File: dao/src/main/scala/za/co/absa/enceladus/dao/auth/MenasCredentials.scala [hunk 2]
```diff
@@ -55,9 +55,9 @@ object MenasKerberosCredentials {
     * @return An instance of Menas Credentials.
     */
   def fromFile(path: String)(implicit spark: SparkSession): MenasKerberosCredentials = {
-    val fsUtils = new FileSystemVersionUtils(spark.sparkContext.hadoopConfiguration)
+    val fsUtils = new HdfsUtils(spark.sparkContext.hadoopConfiguration)
 
-    val localKeyTabPath = fsUtils.getLocalPathToFile(path)
+    val localKeyTabPath = fsUtils.getLocalPathToFileOrCopyToLocal(path)
```
Comment (dk1844) [inline dao/src/main/scala/za/co/absa/enceladus/dao/auth/MenasCredentials.scala:60]: The original logic is to load from local path and fallback on the same path to HDFS. Do we want to have the option to fallback to a menas-credentials file on S3?

File: utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+   *
+   * @param location        s3location - bucket & path are used
+   * @param accumulateOp    operation to accumulate
+   * @param initialAccValue (initial/carry-over) accumulator value
+   * @param breakOut        allows to break the recursion prematurely when the defined value equals the currently accumulated value.
+   *                        Default: None = no break out
+   * @tparam T accumulator value type
+   * @return accumulated value
+   */
+  private def listAndAccumulateRecursively[T](location: S3Location,
```
Comment (dk1844) [inline utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala:235]: All List-and-do-something methods are using this helper in order to support pagination > 1000 items. The related test suite attempts to test this property with an optional breakOut (early recursion return).

File: utils/src/main/scala/za/co/absa/enceladus/utils/fs/example/S3FsUtilsTestJob.scala [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.spark.sql.SparkSession
+import org.slf4j.LoggerFactory
+import software.amazon.awssdk.regions.Region
+import za.co.absa.atum.persistence.S3KmsSettings
+import za.co.absa.atum.utils.S3Utils
+import za.co.absa.enceladus.utils.fs.S3FsUtils
+
+// todo remove or create a integtest like this instead.
+// implementation is directly suited to be runnable locally with a saml profile.
+object S3FsUtilsTestJob {
```
Comment (dk1844) [inline utils/src/main/scala/za/co/absa/enceladus/utils/fs/example/S3FsUtilsTestJob.scala:27]: This entrypoint is entirely designed for testrunning the S3FsUtils on local machine against a real s3 using the local saml profile. Future integTest can be heavily inspired by this class, so maybe it could stay until then?

File: dao/src/main/scala/za/co/absa/enceladus/dao/auth/MenasCredentials.scala (hunk 2 above)
Comment (Zejnilovic) [reply dao/src/main/scala/za/co/absa/enceladus/dao/auth/MenasCredentials.scala:60]: Are you asking if we want the ability to read menas-credentials file from S3 in general? Answer is yes

Or in some special case?

File: plugins-builtin/src/test/scala/za/co/absa/enceladus/plugins/buildin/kafka/ControlInfoSerSuite.scala [hunk 5]
```diff
@@ -16,11 +16,11 @@
 package za.co.absa.enceladus.plugins.buildin.kafka
 
 import org.apache.commons.io.IOUtils
-import org.scalatest.FunSuite
+import org.scalatest.funsuite.AnyFunSuite
```
Comment (Zejnilovic) [inline plugins-builtin/src/test/scala/za/co/absa/enceladus/plugins/buildin/kafka/ControlInfoSerSuite.scala:19]: Can I ask why these changes?

File: dao/src/main/scala/za/co/absa/enceladus/dao/auth/MenasCredentials.scala (hunk 2 above)
Comment (dk1844) [reply dao/src/main/scala/za/co/absa/enceladus/dao/auth/MenasCredentials.scala:60]: Thanks, yes, the former (`ability to read menas-credentials file from S3 in general`).

File: plugins-builtin/src/test/scala/za/co/absa/enceladus/plugins/buildin/kafka/ControlInfoSerSuite.scala (hunk 5 above)
Comment (dk1844) [reply plugins-builtin/src/test/scala/za/co/absa/enceladus/plugins/buildin/kafka/ControlInfoSerSuite.scala:19]: This is due to scala test version update.

https://www.scalatest.org/release_notes/3.1.0#scalatest_deprecations

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala [hunk 6]
```diff
@@ -169,84 +176,93 @@ trait StandardizationExecution extends CommonJobExecution {
...
+      spark,
+      "std",
+      preparationResult.pathCfg.rawPath,
+      preparationResult.pathCfg.standardizationPath,
+      menasCredentials.username,
+      args.mkString(" ")
+    )
+
+    cmd.rowTag.foreach(rowTag => Atum.setAdditionalInfo("xml_row_tag" -> rowTag))
+    cmd.csvDelimiter.foreach(delimiter => Atum.setAdditionalInfo("csv_delimiter" -> delimiter))
```
Comment (Zejnilovic) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala:216]: Why do we only provide these 2 as additional info? This might a mistake even in the master branch.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/common/ControlInfoValidationSuite.scala [hunk 7]
```diff
@@ -15,39 +15,41 @@
...
-class ControlInfoValidationSuite extends FunSuite {
+class ControlInfoValidationSuite extends AnyFunSuite {
   import za.co.absa.atum.core.Constants._
 
   private val checkpoints1 = List(
     Checkpoint("raw", None, None, "", "", "", 0, List(
-      Measurement("", controlTypeAbsAggregatedTotal, "", 0),
-      Measurement("", controlTypeRecordCount, "", 11)
+      Measurement("", ControlType.AbsAggregatedTotal.value, "", 0),
+      Measurement("",  ControlType.Count.value, "", 11)
```
Comment (Zejnilovic) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/common/ControlInfoValidationSuite.scala:30]:
```suggestion
      Measurement("", ControlType.Count.value, "", 11)
```

File: utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+   * Deletes a distributed-FS directory and all its contents recursively
+   */
+  override def deleteDirectoryRecursively(distPath: String): Unit = {
+
+    // setup accumulation
+    val location = distPath.toS3Location(region)
+
+    def accumulateSizeOp(acc: Unit, response: ListObjectsV2Response): Unit = { // side-effect, "accumulates" to unit
+      val objects = response.contents().asScala
+      if (objects.size > 0) {
```
Comment (Zejnilovic) [inline utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala:154]:
```suggestion
      if (objects.nonEmpty) {
```

Comment (Zejnilovic) [reply utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala:154]: Not the only occurance

File: spark-jobs/src/main/resources/reference.conf [hunk 9]
```diff
@@ -95,3 +95,9 @@ timezone="UTC"
 # Optional security settings
 #kafka.security.protocol="SASL_SSL"
 #kafka.sasl.mechanism="GSSAPI"
+
+# S3 specific settings:
+s3.region = "eu-west-1" # can be set here in general
+# recommended to set externally
```
[the remark this answers is not in the document]
Comment (dk1844) [reply spark-jobs/src/main/resources/reference.conf:101]: `s3.region` can be in the `reference.conf` (overridable default), `s3.kmsKeyId` is expected to be only set externally.

I will change the wording to make it clearer.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala (hunk 6 above)
Comment (dk1844) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala:216]: This was in the code before, I haven't changed it, so I don't know.

File: spark-jobs/src/main/resources/reference.conf (hunk 9 above)
[the remark this answers is not in the document]
Comment (dk1844) [reply spark-jobs/src/main/resources/reference.conf]: Done.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/common/ControlInfoValidationSuite.scala (hunk 7 above)
Comment (dk1844) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/common/ControlInfoValidationSuite.scala]: Thanks, fixed.

File: utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala (hunk 8 above)
Comment (dk1844) [reply utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala]: Thanks, fixed.

File: dao/src/main/scala/za/co/absa/enceladus/dao/auth/MenasCredentials.scala (hunk 2 above)
Comment (Zejnilovic) [reply dao/src/main/scala/za/co/absa/enceladus/dao/auth/MenasCredentials.scala:60]: I think fallback is not necessary

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala (hunk 6 above)
Comment (Zejnilovic) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala:216]: I wanted to tag @benedeki and somehow failed.

File: utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala (hunk 8 above)
Comment (Zejnilovic) [reply utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala]: What about `keys.length > 0`

Comment (dk1844) [reply utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala]: True, I searched for `size > 0` 😆 . Fixed.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/ConformanceExecution.scala [hunk 10]
```diff
@@ -43,37 +46,37 @@ import za.co.absa.enceladus.utils.schema.SchemaUtils
...
-    //val stdDirSize = fsUtils.getDirectorySize(preparationResult.pathCfg.standardizationPath)
-    //preparationResult.performance.startMeasurement(stdDirSize) // TODO fix for s3 [ref issue #1416]
+
+    val stdDirSize = fsUtils.getDirectorySize(preparationResult.pathCfg.standardizationPath)
+    preparationResult.performance.startMeasurement(stdDirSize)
 
     log.info(s"standardization path: ${preparationResult.pathCfg.standardizationPath}")
     log.info(s"publish path: ${preparationResult.pathCfg.publishPath}")
 
-    // Enable Control Framework
```
Comment (AdrianOlosutean) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/ConformanceExecution.scala:61]: So this is not needed anymore?

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala (hunk 6 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala:216]: I think the others are added, but these ones are added only if they are set

File: utils/src/main/scala/za/co/absa/enceladus/utils/fs/example/S3FsUtilsTestJob.scala (hunk 4 above)
Comment (AdrianOlosutean) [reply utils/src/main/scala/za/co/absa/enceladus/utils/fs/example/S3FsUtilsTestJob.scala:26]: I would say it can stay, but use different paths in s3

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala (hunk 6 above)
Comment (Zejnilovic) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationExecution.scala:216]: these are the only reader options provided to additional info

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/ConformanceExecution.scala (hunk 10 above)
Comment (dk1844) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/ConformanceExecution.scala:61]: I moved the comment closer to where it makes more sense.

File: utils/src/main/scala/za/co/absa/enceladus/utils/fs/example/S3FsUtilsTestJob.scala (hunk 4 above)
Comment (dk1844) [reply utils/src/main/scala/za/co/absa/enceladus/utils/fs/example/S3FsUtilsTestJob.scala:26]: I have changed the base-path, what is left can reasonably serve as a generic example.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/ConformanceExecution.scala (hunk 10 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/ConformanceExecution.scala:61]: Aa, so this way that import is not needed anymore

**Review by Zejnilovic — APPROVED** [review]
Code reviewed

**Review by AdrianOlosutean — APPROVED** [review]
Code reviewed
